File: js/data/plan.js
// ============================================================
// PERSONALIZATION ENGINE — Rule-based plan generator
// ============================================================
// Inputs: martial arts, level, goals, session time, days, equipment
// Output: weekly schedule (Mon-Sun)
// Simple, predictable, easy to modify.
// ============================================================

import { WORKOUTS } from './content.js';

/**
 * Generate a weekly plan based on user inputs.
 * Returns array of 7 days: each has { day, type, workoutId, completed }
 */
export function generateWeeklyPlan(profile) {
  const {
    martialArts = [],
    level = 'intermediate',
    goals = [],
    sessionDuration = 30,
    daysPerWeek = 4,
    availableDays = ['monday', 'wednesday', 'friday', 'saturday'],
    equipment = [],
  } = profile;

  // Determine which workout categories to use
  const workouts = selectWorkouts(martialArts, goals, equipment, level, sessionDuration);

  // Build the schedule
  const schedule = buildSchedule(workouts, daysPerWeek, availableDays, sessionDuration, martialArts);

  // Mark today's status
  return schedule.map(day => ({
    ...day,
    ...dayStatus(day.day),
  }));
}

/**
 * Select appropriate workouts based on user inputs.
 */
function selectWorkouts(martialArts, goals, equipment, level, duration) {
  const pool = [];

  // Map martial arts to workout categories
  const arts = Array.isArray(martialArts) ? martialArts : [martialArts];

  for (const art of arts) {
    const categoryWorkouts = getWorkoutsForArt(art);
    for (const w of categoryWorkouts) {
      // Filter by duration (within +/- 10 min)
      const durMatch = Math.abs(w.duration - duration) <= 10 || w.duration <= duration;
      // Filter by equipment (no required equipment, or user has it)
      const equipMatch = !w.equipment || w.equipment.every(e => equipment.includes(e));
      if (durMatch && equipMatch) {
        pool.push({ ...w, priority: scoreWorkout(w, art, goals) });
      }
    }
  }

  // Add strength/conditioning as cross-training
  pool.push({ ...WORKOUTS.strength_circuit, priority: 1 });
  pool.push({ ...WORKOUTS.conditioning_session, priority: 1 });

  // Sort by priority, dedupe by id
  const seen = new Set();
  return pool
    .sort((a, b) => b.priority - a.priority)
    .filter(w => {
      if (seen.has(w.id)) return false;
      seen.add(w.id);
      return true;
    });
}

/**
 * Get workouts for a martial art.
 */
function getWorkoutsForArt(art) {
  const map = {
    boxing: ['boxing_footwork', 'boxing_combinations', 'boxing_defense'],
    taekwondo: ['taekwondo_kicking', 'taekwondo_footwork'],
    muay_thai: ['muay_thai_striking'],
    kickboxing: ['kickboxing_combos'],
    wrestling: ['wrestling_takedowns'],
    bjj: ['bjj_fundamentals'],
    karate: ['boxing_combinations'],
    judo: ['wrestling_takedowns'],
    strength: ['strength_circuit', 'conditioning_session'],
    footwork: ['boxing_footwork', 'taekwondo_footwork'],
  };
  const ids = map[art] || [];
  return ids.map(id => WORKOUTS[id]).filter(Boolean);
}

/**
 * Score a workout based on goal alignment.
 */
function scoreWorkout(workout, art, goals) {
  let score = 1;
  const goalArtMap = {
    footwork: ['boxing_footwork', 'taekwondo_footwork', 'boxing_combinations'],
    technique: ['boxing_combinations', 'boxing_defense', 'taekwondo_kicking', 'muay_thai_striking'],
    conditioning: ['conditioning_session', 'strength_circuit'],
    strength: ['strength_circuit'],
    speed: ['boxing_combinations', 'taekwondo_kicking'],
    coordination: ['boxing_combinations', 'taekwondo_kicking'],
    mobility: ['taekwondo_footwork', 'boxing_footwork'],
    fundamentals: ['boxing_footwork', 'taekwondo_footwork', 'bjj_fundamentals', 'boxing_defense'],
  };
  for (const goal of goals) {
    if (goalArtMap[goal]?.includes(workout.id)) {
      score += 2;
    }
  }
  return score;
}

/**
 * Build the actual weekly schedule.
 */
function buildSchedule(workouts, daysPerWeek, availableDays, duration, martialArts) {
  const allDays = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'];
  const dayNames = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

  // Build schedule array
  const schedule = allDays.map((day, i) => ({
    day,
    dayName: dayNames[i],
    type: 'rest',
    workoutId: null,
    workout: null,
    completed: false,
    missed: false,
  }));

  // Place workouts on user's chosen days
  const targetDays = availableDays.slice(0, daysPerWeek);

  // Distribute workouts: alternate martial arts to avoid fatigue
  let workoutIdx = 0;
  for (const day of targetDays) {
    const dayIdx = allDays.indexOf(day);
    if (dayIdx === -1) continue;

    // Pick the next best workout
    let workout = workouts[workoutIdx % workouts.length];
    workoutIdx++;

    // If this is a strength day and we've already used it, skip
    while (
      (workout.id === 'strength_circuit' || workout.id === 'conditioning_session') &&
      schedule.some(d => d.workoutId === workout.id)
    ) {
      workoutIdx++;
      workout = workouts[workoutIdx % workouts.length] || workout;
    }

    schedule[dayIdx] = {
      ...schedule[dayIdx],
      type: 'training',
      workoutId: workout.id,
      workout,
      duration: workout.duration,
    };
  }

  // Make sure we have at least 1 strength day if training days >= 3
  if (daysPerWeek >= 3 && !schedule.some(d => d.workoutId === 'strength_circuit')) {
    const replaceDay = targetDays[Math.floor(targetDays.length / 2)];
    const idx = allDays.indexOf(replaceDay);
    if (idx !== -1) {
      schedule[idx] = {
        ...schedule[idx],
        type: 'training',
        workoutId: 'strength_circuit',
        workout: WORKOUTS.strength_circuit,
        duration: WORKOUTS.strength_circuit.duration,
      };
    }
  }

  return schedule;
}

/**
 * Return day status: today / past / future
 */
function dayStatus(dayKey) {
  const days = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];
  const today = new Date();
  const todayKey = days[today.getDay()];
  if (todayKey === dayKey) return { isToday: true };
  return {};
}

/**
 * Reschedule a missed session to a rest day.
 */
export function rescheduleSession(schedule, fromDay, toDay) {
  const from = schedule.find(d => d.day === fromDay);
  const to = schedule.find(d => d.day === toDay);
  if (!from || !to || to.type !== 'rest') return schedule;

  return schedule.map(d => {
    if (d.day === fromDay) return { ...d, type: 'rest', workoutId: null, workout: null, missed: false };
    if (d.day === toDay) {
      return { ...d, type: 'training', workoutId: from.workoutId, workout: from.workout, duration: from.duration, missed: true };
    }
    return d;
  });
}

/**
 * Find next training day in schedule.
 */
export function nextTrainingDay(schedule) {
  return schedule.find(d => d.type === 'training');
}

/**
 * Find today's training day.
 */
export function todaysTraining(schedule) {
  return schedule.find(d => d.isToday);
}

File: stance_app_BROKEN_UI_BACKUP/lib/data/plan.dart
// ============================================================
// Personalization engine — rule-based weekly plan generator
// Ported from js/data/plan.js
// ============================================================

import '../models/models.dart';
import 'dart:math';
import 'content.dart';

class PlanEngine {
  /// Score how well a workout fits the user's goals, level, time,
  /// martial arts, and available equipment.
  static int scoreWorkout(Workout w, UserProfile profile, Set<String> userEquipment, {Random? random}) {
    int score = 0;

    // Martial art match
    if (profile.martialArts.contains(w.martialArt)) {
      score += 40;
    }

    // Duration match (target is sessionDuration)
    final diff = (w.duration - profile.sessionDuration).abs();
    if (diff <= 5) {
      score += 30;
    } else if (diff <= 10) {
      score += 20;
    } else if (diff <= 15) {
      score += 10;
    }

    // Goal match
    final artTechniques = ContentLibrary.techniquesForArt(w.martialArt);
    if (profile.goals.contains('footwork') && w.title.toLowerCase().contains('footwork')) {
      score += 25;
    }
    if (profile.goals.contains('conditioning') &&
        (w.title.toLowerCase().contains('condition') ||
            w.martialArt == 'strength')) {
      score += 25;
    }
    if (profile.goals.contains('fundamentals') && w.difficulty == 'intermediate') {
      score += 15;
    }
    if (profile.goals.contains('technique') && artTechniques.isNotEmpty) {
      score += 15;
    }
    if (profile.goals.contains('strength') && w.martialArt == 'strength') {
      score += 25;
    }
    if (profile.goals.contains('speed')) {
      if (w.title.toLowerCase().contains('combo') ||
          w.title.toLowerCase().contains('footwork')) {
        score += 10;
      }
    }
    if (profile.goals.contains('competition')) {
      if (w.martialArt != 'footwork' && w.martialArt != 'strength') {
        score += 10;
      }
    }

    // Equipment match
    bool hasAll = true;
    for (final eq in w.equipment) {
      if (!userEquipment.contains(eq) && eq != 'none') {
        hasAll = false;
        break;
      }
    }
    if (hasAll) {
      score += 10;
    } else {
      score -= 100; // Strong penalty for missing required equipment
    }

    // Difficulty match
    if (profile.level == 'beginner' && w.difficulty == 'intermediate') {
      score += 5;
    } else if (profile.level == 'advanced') {
      score += 5;
    }

    // Variety bonus
    if (random != null) {
      score += random.nextInt(10);
    } else {
      score += _hash(w.id) % 10;
    }

    return score;
  }

  /// Pick the best workouts to fill the weekly schedule.
  static List<Workout> selectWorkouts(UserProfile profile, Set<String> userEquipment, {Random? random}) {
    final all = ContentLibrary.workouts;
    final scored = all.map((w) => MapEntry(w, scoreWorkout(w, profile, userEquipment, random: random)))
        .where((e) => e.value > 0)
        .toList()
      ..sort((a, b) => b.value.compareTo(a.value));

    // Try to balance across martial arts
    final picked = <Workout>[];
    final usedArts = <String>{};
    for (final entry in scored) {
      if (picked.length >= profile.daysPerWeek) break;
      final art = entry.key.martialArt;
      if (usedArts.length < profile.martialArts.length && usedArts.contains(art)) {
        continue; // skip if we've already filled one slot per art
      }
      picked.add(entry.key);
      usedArts.add(art);
    }
    // Fill remainder if needed
    for (final entry in scored) {
      if (picked.length >= profile.daysPerWeek) break;
      if (!picked.contains(entry.key)) {
        picked.add(entry.key);
      }
    }
    return picked.take(profile.daysPerWeek).toList();
  }

  /// Build the weekly schedule for the user's available days.
  static List<ScheduledDay> buildSchedule(UserProfile profile, Set<String> userEquipment, {Random? random}) {
    final picked = selectWorkouts(profile, userEquipment, random: random);
    final orderedDays = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'];

    // Determine "today" for highlighting
    final now = DateTime.now();
    final todayIdx = now.weekday - 1; // 0=Mon..6=Sun

    final availableDays = profile.availableDays.toSet();
    final schedule = <ScheduledDay>[];
    int workoutIdx = 0;
    for (int i = 0; i < orderedDays.length; i++) {
      final day = orderedDays[i];
      if (availableDays.contains(day)) {
        if (workoutIdx < picked.length) {
          final w = picked[workoutIdx++];
          schedule.add(ScheduledDay(
            day: day,
            dayName: _dayName(day),
            type: 'training',
            workoutId: w.id,
            workout: w,
            duration: w.duration,
            isToday: i == todayIdx,
          ));
        } else {
          schedule.add(ScheduledDay(
            day: day,
            dayName: _dayName(day),
            type: 'rest',
            isToday: i == todayIdx,
          ));
        }
      } else {
        schedule.add(ScheduledDay(
          day: day,
          dayName: _dayName(day),
          type: 'rest',
          isToday: i == todayIdx,
        ));
      }
    }
    return schedule;
  }

  /// Reschedule missed sessions: shuffle training days so the upcoming days get a workout.
  static List<ScheduledDay> reschedule(
    List<ScheduledDay> schedule,
    UserProfile profile,
    Set<String> userEquipment, {
    Random? random,
  }) {
    final workouts = <Workout>[];
    for (final s in schedule) {
      if (s.type == 'training' && s.workout != null) workouts.add(s.workout!);
    }
    if (workouts.isEmpty) return schedule;

    final newSchedule = buildSchedule(profile, userEquipment, random: random);
    // Replace completed days
    final merged = <ScheduledDay>[];
    for (int i = 0; i < schedule.length; i++) {
      if (schedule[i].completed) {
        merged.add(schedule[i]);
      } else {
        merged.add(newSchedule[i]);
      }
    }
    return merged;
  }

  static String _dayName(String day) {
    const map = {
      'monday': 'Monday',
      'tuesday': 'Tuesday',
      'wednesday': 'Wednesday',
      'thursday': 'Thursday',
      'friday': 'Friday',
      'saturday': 'Saturday',
      'sunday': 'Sunday',
    };
    return map[day] ?? day;
  }

  static int _hash(String s) {
    int h = 0;
    for (int i = 0; i < s.length; i++) {
      h = (h * 31 + s.codeUnitAt(i)) & 0x7fffffff;
    }
    return h;
  }
}
